include/cuco/detail/hash_functions/xxhash.cuh

```diff
@@ -0,0 +1,354 @@
+    constexpr auto nbytes         = sizeof(Key);
+    char const* const bytes       = (char const*)&key;      ///< per-byte access
+    uint32_t const* const blocks4 = (uint32_t const*)&key;  ///< 4-byte word access
+    uint64_t const* const blocks8 = (uint64_t const*)&key;  ///< 8-byte word access
+
+    uint64_t offset = 0;
+    uint64_t h64;
+
+    // data can be processed in 32-byte chunks
+    if constexpr (nbytes >= 32) {
+      constexpr auto limit = nbytes - 32;
+      uint64_t v1          = seed_ + prime1 + prime2;
+      uint64_t v2          = seed_ + prime2;
+      uint64_t v3          = seed_;
+      uint64_t v4          = seed_ - prime1;
+
+      do {
+        // pipeline 4*8byte computations
+        auto const pipeline_offset = offset >> 3;  // optimized division by 8
+        v1 += blocks8[pipeline_offset] * prime2;
+        v1 = rotl(v1, 31);
+        v1 *= prime1;
+        v2 += blocks8[pipeline_offset + 1] * prime2;
+        v2 = rotl(v2, 31);
+        v2 *= prime1;
+        v3 += blocks8[pipeline_offset + 2] * prime2;
+        v3 = rotl(v3, 31);
+        v3 *= prime1;
+        v4 += blocks8[pipeline_offset + 3] * prime2;
+        v4 = rotl(v4, 31);
+        v4 *= prime1;
+        offset += 32;
+      } while (offset <= limit);
+
+      h64 = rotl(v1, 1) + rotl(v2, 7) + rotl(v3, 12) + rotl(v4, 18);
+
+      v1 *= prime2;
+      v1 = rotl(v1, 31);
+      v1 *= prime1;
+      h64 ^= v1;
+      h64 = h64 * prime1 + prime4;
+
+      v2 *= prime2;
+      v2 = rotl(v2, 31);
+      v2 *= prime1;
+      h64 ^= v2;
+      h64 = h64 * prime1 + prime4;
+
+      v3 *= prime2;
+      v3 = rotl(v3, 31);
+      v3 *= prime1;
+      h64 ^= v3;
+      h64 = h64 * prime1 + prime4;
+
+      v4 *= prime2;
+      v4 = rotl(v4, 31);
+      v4 *= prime1;
+      h64 ^= v4;
+      h64 = h64 * prime1 + prime4;
+    } else {
+      h64 = seed_ + prime5;
+    }
+
+    h64 += nbytes;
+
+    // remaining data can be processed in 8-byte chunks
+    if constexpr ((nbytes % 32) >= 8) {
+      for (; offset <= nbytes - 8; offset += 8) {
+        uint64_t k1 = blocks8[offset >> 3] * prime2;
+        k1          = rotl(k1, 31) * prime1;
+        h64 ^= k1;
+        h64 = rotl(h64, 27) * prime1 + prime4;
+      }
+    }
+
+    // remaining data can be processed in 4-byte chunks
+    if constexpr (((nbytes % 32) % 8) >= 4) {
+      for (; offset <= nbytes - 4; offset += 4) {
+        h64 ^= (blocks4[offset >> 2] & 0xFFFFFFFFULL) * prime1;
+        h64 = rotl(h64, 23) * prime2 + prime3;
+      }
+    }
+
+    // the following loop is only needed if the size of the key is no multiple of a previous block
+    // size
+    if constexpr (nbytes % 4) {
+      while (offset < nbytes) {
+        h64 += (bytes[offset] & 0xFF) * prime5;
+        h64 = rotl(h64, 11) * prime1;
+        ++offset;
+      }
+    }
+    return finalize(h64);
+  }
+
+ private:
+  constexpr __host__ __device__ uint64_t rotl(uint64_t h, int8_t r) const noexcept
```

Comment:
worth trying `__force_inline__` since the operation is constantly repeated.

Reply:
I was thinking of creating a base class for xxhash for common utilities like `rotl` and `finalize` then realized it doesn't provide much convenience.

Reply:
Just tested this: No noticeable change in the benchmark results with and without `__forceinline__`.